### PR TITLE
Added error message to impossible instanceof rule to warn when comparing a trait.

### PR DESCRIPTION
**Pull request by mad-briller**
implements https://github.com/phpstan/phpstan/issues/7720#issuecomment-1200126569

### CODE REVIEW

File: tests/PHPStan/Type/ObjectTypeTest.php [hunk 1]
```diff
@@ -418,6 +419,16 @@ public function dataIsSuperTypeOf(): array
 				new ObjectType(ExtendsThrowable::class),
 				TrinaryLogic::createMaybe(),
 			],
+			59 => [
+				new ObjectType(DateTime::class),
+				new ObjectType(ConstantNumericComparisonTypeTrait::class),
```
Comment (mad-briller) [inline tests/PHPStan/Type/ObjectTypeTest.php:424]: pretty sure php has no built-in traits so i used one that phpstan has, i feel like theres a better way to do this though

File: src/Rules/Classes/ImpossibleInstanceOfRule.php [hunk 2]
```diff
@@ -57,6 +60,20 @@ public function processNode(Node $node, Scope $scope): array
 			}
 		}
 
+		if ($className !== null) {
+			$classReflection = $this->reflectionProvider->getClass($className);
```
Comment (ondrejmirtes) [inline src/Rules/Classes/ImpossibleInstanceOfRule.php]: You first need to ask about hasClass, only then it's safe to call getClass.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by ondrejmirtes** [comment]
Maybe this rule is a better fit for this problem? https://github.com/phpstan/phpstan-src/blob/1.8.x/src/Rules/Classes/ExistingClassInInstanceOfRule.php

File: src/Rules/Classes/ImpossibleInstanceOfRule.php (hunk 2 above)
Comment (mad-briller) [reply src/Rules/Classes/ImpossibleInstanceOfRule.php]: ah of course, thanks

**Comment by mad-briller** [comment]
i think either location works, though i think "existing class" is a bit more indirect than "impossible instanceof" when it comes to comparing a trait in instanceof as a trait is not a class
happy to move if it you think the other location is better!

**Comment by ondrejmirtes** [comment]
On the other hand there's already ReflectionProvider in the "existing class" rule, and to me it makes more sense there because trait is not an existing class 😊 I just forgot that rule exists, sorry 😊

**Comment by mad-briller** [comment]
no worries mate, i've moved the check to the ExistingClassInInstanceOf rule

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**